File: packages/javascript/lib/node/file/getFileHashFromStream.ts
import { createHash } from 'crypto';
import { pipeline } from 'stream/promises';

export async function getFileHashFromStream(
	fileStream: NodeJS.ReadableStream,
	algorithm: 'md5' | 'sha1' | 'sha256' | 'sha512' = 'md5'
): Promise<string> {
	if (!fileStream) {
		throw new Error('Invalid file stream detected');
	}

	const hashHandler = createHash(algorithm);

	await pipeline(fileStream, hashHandler);

	return hashHandler.digest('hex');
}

File: packages/python/qsu/file/getFileHashFromStream.py
import hashlib


def getFileHashFromStream(fileStream, algorithm: str = 'md5') -> str:
	if not fileStream:
		raise Exception('Invalid file stream detected')

	hashHandler = hashlib.new(algorithm)

	while True:
		chunk = fileStream.read(65536)

		if not chunk:
			break

		if isinstance(chunk, str):
			chunk = chunk.encode('utf-8')

		hashHandler.update(chunk)

	return hashHandler.hexdigest()
